### PR TITLE
Jormun: geovelo's call to direct path

### PR DESCRIPTION
**Pull request by pbougue**
TODO:

- [x] Warn when direct_path != isochrone (one more call to API, maybe in debug) ? 
> Not so usefull for now

- [x] Attach "correctly" fallback to tc (experimental, in another PR) ? 
> Done in this PR (hand-tested with geovelo, a bigger refacto is needed on experimental)

### CODE REVIEW

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 1]
```diff
@@ -52,16 +52,33 @@ def __init__(self, instance, service_url, id='geovelo', timeout=10, api_key=None
...
+    def _make_request_arguments_direct_path(cls, origin, destination):
+        coord_orig = get_pt_object_coord(origin)
+        coord_dest = get_pt_object_coord(destination)
+        return {
+            'waypoints': [
+                {'latitude': coord_orig.lat, 'longitude': coord_orig.lon},
+                {'latitude': coord_dest.lat, 'longitude': coord_dest.lon}
+            ],
+            'bikeDetails': {
+                'profile': 'BEGINNER',
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: open question, are there different interesting profiles ? shouldn't this be a connector configuration ?

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 2]
```diff
@@ -125,69 +142,98 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
             raise InvalidArguments('Geovelo, managing only 1-n in connector, requested {}-{}'
                                    .format(len(origins), len(destinations)))
 
-        data = self._make_request_arguments(origins, destinations)
+        data = Geovelo._make_request_arguments_isochrone(origins, destinations)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: you can use `cls` here

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 3]
```diff
@@ -125,69 +142,98 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
...
         r = self._call_geovelo('{}/{}'.format(self.service_url, 'api/v2/routes_m2m'),
                                requests.post, json.dumps(data))
-        self._check_response(r)
+        Geovelo._check_response(r)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: same here

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 4]
```diff
@@ -125,69 +142,98 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
...
         resp_json = r.json()
 
         if len(resp_json) - 1 != len(origins) * len(destinations):
             logging.getLogger(__name__).error('Geovelo nb response != nb requested')
             raise UnableToParse('Geovelo nb response != nb requested')
 
-        return self._get_matrix(resp_json)
+        return Geovelo._get_matrix(resp_json)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: you can use `cls` here

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 5]
```diff
@@ -196,14 +242,20 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, datetime, c
             logging.getLogger(__name__).error('Geovelo, mode {} not implemented'.format(mode))
             raise InvalidArguments('Geovelo, mode {} not implemented'.format(mode))
 
-        data = Geovelo._make_request_arguments([pt_object_origin], [pt_object_destination])
-        r = self._call_geovelo('{}/{}'.format(self.service_url, 'api/v2/routes_m2m'),
+        data = Geovelo._make_request_arguments_direct_path(pt_object_origin, pt_object_destination)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: you can use `self._make...`

File: source/jormungandr/jormungandr/street_network/geovelo.py (hunk 2 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: data = ~~Geovelo~~cls._make_....

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 6]
```diff
@@ -196,14 +242,20 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, datetime, c
...
+        r = self._call_geovelo('{}/{}'.format(self.service_url, 'api/v2/computedroutes?'
+                                                                'instructions=true&'
+                                                                'elevations=false&'
+                                                                'geometry=true&'
+                                                                'single_result=true&'
+                                                                'bike_stations=false&'
+                                                                'objects_as_ids=true&'),
                                requests.post, json.dumps(data))
-        self._check_response(r)
+        Geovelo._check_response(r)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: you can use `self`

File: source/jormungandr/jormungandr/street_network/geovelo.py (hunk 3 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: ~~Geovelo~~cls._check_....

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 7]
```diff
@@ -196,14 +242,20 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, datetime, c
...
         resp_json = r.json()
 
-        if len(resp_json) != 2:
+        if len(resp_json) != 1:
             logging.getLogger(__name__).error('Geovelo nb response != nb requested')
             raise UnableToParse('Geovelo nb response != nb requested')
 
-        return self._get_response(resp_json, mode, pt_object_origin, pt_object_destination, datetime, clockwise)
+        return Geovelo._get_response(resp_json, mode, pt_object_origin, pt_object_destination, datetime, clockwise)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: you can use `self`

File: source/jormungandr/jormungandr/street_network/valhalla.py [hunk 8]
```diff
@@ -140,36 +112,40 @@ def _get_response(self, json_resp, mode, pt_object_origin, pt_object_destination
...
-        for leg in json_resp['trip']['legs']:
-            journey = resp.journeys.add()
-            journey.duration = leg['summary']['time']
-            if clockwise:
-                journey.departure_date_time = datetime
-                journey.arrival_date_time = datetime + journey.duration
-            else:
-                journey.departure_date_time = datetime - journey.duration
-                journey.arrival_date_time = datetime
+        journey = resp.journeys.add()
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/valhalla.py:118]: I don't understand why you changed this ?

File: source/jormungandr/jormungandr/street_network/valhalla.py [hunk 9]
```diff
@@ -201,10 +178,10 @@ def _get_valhalla_mode(self, kraken_mode):
 
     def _make_request_arguments(self, mode, pt_object_origin, pt_object_destinations, request, api='route', max_duration=None):
 
-        valhalla_mode = self._get_valhalla_mode(mode)
-        destinations = [self._format_coord(destination, api) for destination in pt_object_destinations]
+        valhalla_mode = Valhalla._get_valhalla_mode(mode)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/valhalla.py]: you can keep `self` there

File: source/jormungandr/jormungandr/street_network/geovelo.py (hunk 1 above)
Comment (pbougue) [reply source/jormungandr/jormungandr/street_network/geovelo.py]: I add a comment on possible values.
It seems too soon to put that in the connector or at call (following profile and bike speed?). In my opinion, we have to wait until geovelo manages parameters consistently in isochrone and in direct_path.

File: source/jormungandr/jormungandr/street_network/valhalla.py (hunk 8 above)
Comment (pbougue) [reply source/jormungandr/jormungandr/street_network/valhalla.py:118]: We used to add one journey per leg (which is a section).
It was invisible as we are supposed to have only one leg, but it was not consistent. I did the cleanup along the comparison while I tried to use pretty much the same thing in geovelo.

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 10]
```diff
@@ -52,16 +52,33 @@ def __init__(self, instance, service_url, id='geovelo', timeout=10, api_key=None
...
+        coord_dest = get_pt_object_coord(destination)
+        return {
+            'waypoints': [
+                {'latitude': coord_orig.lat, 'longitude': coord_orig.lon},
+                {'latitude': coord_dest.lat, 'longitude': coord_dest.lon}
+            ],
+            'bikeDetails': {
+                'profile': 'BEGINNER', # can be MEDIAN, EXPERT
+                'bikeType': 'TRADITIONAL', # can be 'BSS'
+                # 'averageSpeed': '6' # in km/h, BEGINER sets it to 13
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/street_network/geovelo.py:79]: shouldn't we use the query parameters there ?

Comment (TeXitoi) [reply source/jormungandr/jormungandr/street_network/geovelo.py:79]: when we can do it on the matrix call, but not before

Comment (pbougue) [reply source/jormungandr/jormungandr/street_network/geovelo.py:79]: To elaborate a bit: yes we should, but there would be even more difference between isochrone and directpath.
As geovelo's parameters may change when they will implem speed for isochrone, and also as we may change a bit experimental when refactoring it (even though this part shouldn't change much)... It's does not seem worth doing it now.

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 11]
```diff
@@ -78,46 +78,57 @@ def get_max_fallback_duration(request, mode, dp_durations={}):
...
 
 
-def make_direct_path_key(dep_mode, orig_uri, dest_uri, datetime, clockwise, reverse_sections):
-    # datetime is not taken into consideration because we assume that
-    # a direct path from A to B remains the same even the departure time are different
-    return dep_mode, orig_uri, dest_uri, clockwise, reverse_sections
+def make_direct_path_key(dep_mode, orig_uri, dest_uri, is_fallback_at_end,
+                         datetime, is_start_fallback_datetime):
+    '''
+    A, B and mode matters obviously
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: A and B does not appear above, it's a bit confusing

and maybe add a bit more general comment first

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 12]
```diff
@@ -391,56 +402,52 @@ def get_update_crowfly_duration_futures(self):
         return origin_futures, destination_futures
 
     @staticmethod
-    def _get_direct_path(instance, mode, pt_object_origin, pt_object_destination, datetime, clockwise,
-                         request, reverse_sections=False):
+    def _get_direct_path(instance, mode, pt_object_origin, pt_object_destination,
+                         datetime, is_start_fallback_datetime, request, is_fallback_at_end):
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: what is is_start_fallback_datetime ?

File: source/jormungandr/jormungandr/street_network/valhalla.py (hunk 8 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/street_network/valhalla.py:118]: @antoine-de he pulled out some one-shot code from the loop

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 13]
```diff
@@ -42,11 +42,11 @@
 import collections
 import copy
 
-def create_crowfly(pt_journey, _from, to, begin, end, mode='walking'):
+def create_crowfly(pt_journey, crowfly_origin, crowfly_destination, begin, end, mode='walking'):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/experimental.py:45]: 👍

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 14]
```diff
@@ -466,63 +473,63 @@ def worker_journey(dep_mode=dep_mode,
...
+                        datetime, is_start_fallback_datetime,
+                        fallback_direct_path_pool, is_fallback_at_end):
+        fallback_dp = get_direct_path_if_exists(fallback_direct_path_pool, mode, pb_from.uri, pb_to.uri,
+                                                datetime, is_start_fallback_datetime, is_fallback_at_end)
+        fallback_copy = copy.deepcopy(fallback_dp)
+        pt_journey.duration += fallback_copy.journeys[0].duration
+        pt_journey.durations.walking += fallback_copy.journeys[0].durations.walking
+        _extend_pt_sections_with_direct_path(pt_journey, fallback_copy)
+
+    def _build_from(self, journey, first_section, journey_origin, crowfly_stop_points, odt_stop_points,
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: so we may rename `journey` to `pt_journey`?

**Review by antoine-de — APPROVED** [review]

**Pull request merged**